### PR TITLE
Harden Agent approval and cancellation boundaries

### PR DESCRIPTION
**Pull request by ross0x01**
## Summary
- harden reusable Agent approval grants by rejecting dynamic Windows shell syntax/wrappers and requiring one-time approval for file write, append, and edit operations
- bind each successful Agent approval to the exact Desktop connection or E2B environment checked by the worker, rejecting a replacement connection before terminal or file execution
- bind local/Desktop cancellation success to an explicit native termination acknowledgement while retaining uncertain sessions for retry
- limit Enter approval to the focused Allow control, remove raw approval targets from Trigger telemetry, and serialize log-only Durable Agent errors generically to clients

## Root causes and user impact
- reusable command grants classified the scanned command text without conservatively rejecting Windows environment expansion and shell-wrapper equivalents
- reusable file grants compared lexical paths even though E2B, Local, and Desktop backends cannot bind that string to the same canonical filesystem object across the later sink operation
- the worker rechecked Desktop A after approval but returned no sandbox identity; terminal and file tools independently reacquired the current sandbox, so a precisely timed same-account replacement with Desktop B could receive the approved operation
- approval logs and run metadata persisted raw target or prefix material
- local cancellation treated publish/request success as process termination and could discard command tracking after a false or missing acknowledgement
- a document-wide Enter listener approved from body or unrelated focus
- log-only ChatSDKError diagnostics were serialized into the client stream

The Desktop replacement race is calibrated as Medium at most, not High: it is same-account and requires a narrow reconnect/replacement interval. It is still a real boundary violation because the repository explicitly scopes reusable approvals to `connection:<connectionId>`. The file compatibility tradeoff is intentional: file write, append, and edit now require fresh approval because the current cross-backend contract cannot prove canonical object identity through symlinks, junctions, and TOCTOU windows.

## Validation
- `corepack pnpm jest app/components/ChatInput/__tests__/AgentApprovalPrompt.test.tsx app/services/__tests__/desktop-sandbox-bridge.test.ts lib/__tests__/errors.test.ts lib/ai/tools/__tests__/run-terminal-cmd.test.ts lib/ai/tools/__tests__/interact-terminal-session.test.ts lib/ai/tools/__tests__/file.test.ts lib/ai/tools/utils/__tests__/centrifugo-sandbox.test.ts lib/ai/tools/utils/__tests__/hybrid-sandbox-manager.test.ts lib/api/__tests__/agent-long-contracts.test.ts lib/chat/__tests__/agent-approval-grants.test.ts packages/local/src/__tests__/command-cancellation.test.ts trigger/__tests__/agent-long-post-wait-authorization.test.ts --runInBand --silent` — 12 suites, 306 tests
- `corepack pnpm jest app/components/ChatInput/__tests__ app/services/__tests__ lib/__tests__ lib/ai/tools/__tests__ lib/ai/tools/utils/__tests__ lib/api/__tests__ lib/chat/__tests__ packages/local/src/__tests__ trigger/__tests__ --runInBand --silent` — 86 suites, 1,085 tests
- `corepack pnpm typecheck`
- `corepack pnpm lint` — 0 errors; 6 pre-existing warnings
- `corepack pnpm --dir packages/local build`
- changed TypeScript Prettier check and `git diff --check`

Exploit reruns cover COMSPEC case variants, SystemRoot/windir wrappers, caret-obfuscated cmd, cmd/cmd.exe, command.com, call, start, WSL, and quoted/unquoted Windows paths; direct `ping` prefix reuse remains valid. Separate file write/append/edit tests reject both new and legacy reusable grants. Cancellation tests cover publish rejection, native false, missing acknowledgement, retained tracking/retry, and confirmed success. Error tests prove generic client text with intact server diagnostics. Desktop binding tests prove terminal execution, interactive PTY creation, and file write approved on A cannot reach B, while same-connection execution remains valid.

## Remaining validation gaps
- Windows runtime classification and Desktop process-tree behavior need a Windows Local/Desktop smoke test. All platform-independent parser/grant tests pass.
- The connection-race regression uses in-memory sandbox-manager doubles at the real tool acquisition boundary; no live authenticated two-Desktop reconnect timing test touched Convex or Centrifugo resources.
- Rust tooling is unavailable in this worktree, so the Tauri Rust changes were reviewed but not compiled locally.
- Chrome loaded the local app without console errors, but an authenticated live Agent approval card was unavailable without touching live Trigger resources. The component interaction tests cover focused Allow Enter and body/unrelated-focus rejection.

## Manual verification
1. **Windows Local/Desktop:** approve a direct reusable command such as `ping`, confirm another matching direct command reuses it, then confirm `%COMSPEC%`, case variants, environment-root cmd paths, cmd/cmd.exe, PowerShell, WSL, and wrapper variants require fresh approval.
2. **Desktop replacement:** begin approval on Desktop A, approve it, then replace/reconnect as Desktop B before execution. Confirm the operation is rejected with retry guidance and does not execute on B; retry and approve for B, then confirm it executes there.
3. **File grants:** approve write, append, and edit once each; confirm the UI offers no reusable scope and the next operation asks again. Exercise a symlink/junction plus `..\\` destination and confirm it cannot reuse an older file grant.
4. **Cancellation:** start a long-lived child tree in Local and Desktop, cancel it, and confirm success appears only after the child tree exits. Simulate native/transport failure and confirm an uncertain failure plus retained resumable session, then retry termination.
5. **Agent approval keyboard:** in Google Chrome, open an authenticated approval card. Press Enter with body/unrelated focus (no approval), then focus **Allow once** and press Enter (one approval); verify click and keyboard accessibility.
6. **Durable error display:** trigger a safe test error with log-only diagnostics; confirm the client shows generic text while bounded server logs retain the internal diagnostic and the client payload contains no operation/request/failure/chat/user metadata.

No production service, live Trigger resource, billing provider, real user data, or public target was used.

## Summary by CodeRabbit

* **New Features**
  * Added acknowledgement-based command cancellation results across local, desktop, and sandbox flows.
  * Introduced a sandbox cancellation “readiness” hook for coordinated abort/confirmation.
  * Added sandbox identity to successful tool approvals to keep subsequent tool execution consistent.

* **Bug Fixes**
  * Prevented approval submission when Enter is pressed outside the focused approval control.
  * Tightened terminal/file approval validation (including Windows shell variants and file-operation freshness).
  * Fixed cancellation result publication so the desktop side always responds.

* **Security**
  * Reduced streamed error detail and reshaped approval telemetry to avoid exposing sensitive diagnostics/targets.

* **Tests**
  * Expanded cancellation, keyboard/approval, and serialization regression coverage with new edge cases.

### CODE REVIEW

File: packages/local/src/command-cancellation.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      proc.off("close", handleTerminal);
+      proc.off("error", handleTerminal);
+      resolve(confirmed);
+    };
+    const handleTerminal = () => finish(true);
+    const timeoutId = setTimeout(() => finish(false), timeoutMs);
+    timeoutId.unref?.();
+
+    proc.once("close", handleTerminal);
+    proc.once("error", handleTerminal);
```
[the remark this answers is not in the document]
Comment (ross0x01) [reply packages/local/src/command-cancellation.ts]: Fixed in bc73b73d. `close` now confirms termination, while `error` returns false and keeps the cancellation uncertain. Added an emitted-error regression test.

**Comment by ross0x01** [comment]
Addressed the remaining CodeRabbit review-body suggestions in bc73b73d: the acknowledgement deadline now starts before cancellation publish and stale publish outcomes are attempt-guarded; approval telemetry tests now inspect each event payload's exact keys. Added stalled-publish/late-rejection coverage and reran 9 focused suites (259 tests), 84 owning suites (1,071 tests), typecheck, lint, formatting, and the local runner build.

**Comment by ross0x01** [comment]
Reviewed the refreshed CodeRabbit result for a0e6d5be: there are no new actionable review threads. The docstring-coverage item is a non-blocking style heuristic, not a functional or security defect in this patch; adding broad unrelated boilerplate would expand the scoped security diff, so no change is needed. The new Desktop A-to-B regressions and same-connection positive control are covered by the 306-test focused run and 1,085-test owning-suite run.
